### PR TITLE
[AIRFLOW-XXXX] Add InProcessExecutor

### PR DESCRIPTION
**Pull request by turbaszek**
Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-XXX
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [ ] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

File: airflow/executors/python_executor.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+import logging
+import subprocess
+import threading
+
+from airflow import AirflowException
+from airflow.executors.base_executor import BaseExecutor
+from airflow.utils.db import provide_session
+from airflow.utils.state import State
+
+logger = logging.getLogger(__name__)
```
Comment (turbaszek) [inline airflow/executors/python_executor.py]:
```suggestion

```
Is there any reason for using external logger? `BaseExecutor` inherits from`LogginMixin` thus logging can be done in executor via `self.log.info(msg)`.

File: airflow/executors/python_executor.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+
+class PythonExecutor(BaseExecutor):
```
Comment (turbaszek) [inline airflow/executors/python_executor.py]: I think the name is quite misleading (all executors are python) what about the old name `InProcess`? Or `SingleProcess`?

File: airflow/executors/python_executor.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+                self.log.error("Failed to execute task: %s.", str(e))
+            except KeyboardInterrupt as e:
+                task_failed = True
+                fail_fast = True
+                self.change_state(key, State.FAILED)
+                self.log.exception("Interrupted to execute task: %s.", str(e))
+            except Exception as e:
+                task_failed = True
+                self.change_state(key, State.FAILED)
+                self.log.exception("Failed to execute task: %s.", str(e))
```
Comment (turbaszek) [inline airflow/executors/python_executor.py]: I would extract this to other method like `run_task` that would set `task_failed = self.run_task()` then the loop would be more readable imho.

File: airflow/executors/python_executor.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+            (
+                task_instance,
+                dict(
+                    mark_success=mark_success,
+                    # ignore_all_deps =ignore_all_deps,
+                    # ignore_depends_on_past=ignore_depends_on_past,
+                    # ignore_task_deps=ignore_task_deps,
+                    # ignore_ti_state=ignore_ti_state,
+                    pool=pool,
+                    # pickle_id=pickle_id,
```
Comment (turbaszek) [inline airflow/executors/python_executor.py]: Should we remove this comment together with unused params?

File: airflow/executors/python_executor.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    # EOF override
+
+    @provide_session
+    def _run_task_instance(self, ti, mark_success, pool, session=None):
+        # set proper state and try number to keep logger in sync
+        ti.state = State.RUNNING
+        ti._try_number += 1
+        # let save state
+        session.merge(ti)
+        session.commit()
```
Comment (turbaszek) [inline airflow/executors/python_executor.py]: We should use `TaskInstance.set_state` for that

File: airflow/executors/python_executor.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        from airflow.utils.log import logging_mixin
+
+        logging_mixin.set_context(logging.root, ti)
+
+        try:
+            ti._run_raw_task(mark_success=mark_success, job_id=ti.job_id, pool=pool)
+        finally:
+            for handler in logging.root.handlers:
+                if handler.name == "task":
+                    handler.close()
```
Comment (turbaszek) [inline airflow/executors/python_executor.py]: What is the reason behind setting context?

File: airflow/executors/python_executor.py (hunk 2 above)
Comment (evgenyshulman) [reply airflow/executors/python_executor.py]: I would vote for InProcess ( our initial name)

File: airflow/executors/python_executor.py (hunk 3 above)
Comment (evgenyshulman) [reply airflow/executors/python_executor.py]: Let just make sure it similar to what we have in other executors)

File: airflow/executors/python_executor.py (hunk 4 above)
Comment (evgenyshulman) [reply airflow/executors/python_executor.py]: Check original Sequential, probably we need to uncomment

File: airflow/executors/python_executor.py (hunk 6 above)
Comment (evgenyshulman) [reply airflow/executors/python_executor.py]: Initially we used our own logging system instead of what airflow provides, now we can just remove this lines

File: airflow/executors/python_executor.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+        task_failed = False
+        for key, command in self.tasks_to_run:
+            ti = command[0]
+            if self.fail_fast and task_failed:
+                logger.info("Setting %s to %s", key, State.UPSTREAM_FAILED)
+                ti.set_state(State.UPSTREAM_FAILED)
+                self.change_state(key, State.UPSTREAM_FAILED)
+                continue
+
+            if self._is_terminated:
```
Comment (turbaszek) [inline airflow/executors/python_executor.py]: Ouch... this statement is always true should be `self._is_terminated.is_set()`

**Comment by turbaszek** [comment]
@potiuk I wait for your awesome PR 🙏🏽
```
airflow/utils/log/file_task_handler.py:1:0: R0401: Cyclic import (airflow -> airflow.executors -> airflow.executors.inprocess_executor) (cyclic-import)
```

**Comment by potiuk** [comment]
Yeah. It just waits for @ashb/@mik-laj final review. And it fails from time to time with 'on_kill" test which waits for your awesome PR to be merged :).

File: airflow/models/taskinstance.py [hunk 8]
```diff
@@ -883,6 +885,12 @@ def _run_raw_task(
                 context = self.get_template_context()
 
                 task_copy = copy.copy(task)
+
+                if issubclass(type(task_copy), BaseSensorOperator) and \
+                        conf.get('core', 'executor') in ("InProcessExecutor", "SequentialExecutor"):
+                    self.log.warning("InProcessExecutor changes sensor mode to 'reschedule'.")
+                    task_copy.mode = 'reschedule'
+
```
Comment (turbaszek) [inline airflow/models/taskinstance.py:909]: This change allow us to run DAGs with sensors using `InProcessExecutor` and `SequentialExecutor`. To be honest I think we can consider replacing sequential with inprocess. 

@potiuk @mik-laj @evgenyshulman any thoughts about that?

**Comment by turbaszek** [comment]
Present state:
We have working executor that can be simply run from IDE. To do this you have to 
- adjust your IDE run configuration (`AIRFLOW__CORE__EXECUTOR=InProcessExecutor` + everything you may need like GOOGLE_DEFAULT_CREDENTIALS)
- add in your DAG file:
```python
if __name__ == '__main__':
    example_dag.clear(reset_dag_runs=True)
    example_dag.run()
```

Possible next steps:
- documentation about using new executor
- InteliJ plugin

File: airflow/config_templates/default_airflow.cfg [hunk 9]
```diff
@@ -91,6 +91,10 @@ default_timezone = utc
 # SequentialExecutor, LocalExecutor, CeleryExecutor, DaskExecutor, KubernetesExecutor
 executor = SequentialExecutor
 
+# Used only with InProcessExecutor. If set to True DAG will fail with first
+# failed task. Helpful for debugging purposes.
+fail_fast = False
```
Comment (potiuk) [inline airflow/config_templates/default_airflow.cfg]: I think for now this should be a conf in "InProcessExecutor" section for now rather than generic conf. Later when/if we introduce fail fast globally we can make it global conf.

File: airflow/executors/base_executor.py [hunk 10]
```diff
@@ -85,7 +85,7 @@ def queue_task_instance(
             ignore_task_deps: bool = False,
             ignore_ti_state: bool = False,
             pool: Optional[str] = None,
-            cfg_path: Optional[str] = None):
+            cfg_path: Optional[str] = None) -> None:
```
Comment (potiuk) [inline airflow/executors/base_executor.py:88]: +1

File: airflow/executors/inprocess_executor.py [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+from airflow import conf
+from airflow.executors.base_executor import BaseExecutor
+from airflow.models.taskinstance import TaskInstance, TaskInstanceKeyType
+from airflow.utils.state import State
+
+
+class InProcessExecutor(BaseExecutor):
+    """
+    This executor is meant for debugging purposes. It is also can
+    be used with sqlite since sqlite.
```
Comment (potiuk) [inline airflow/executors/inprocess_executor.py]: Not finished sentence.

File: airflow/executors/inprocess_executor.py [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+        super().__init__()
+        self.tasks_to_run: List[TaskInstance] = []
+        # Place where we keep information for task instance raw run
+        self.tasks_params: Dict[TaskInstanceKeyType, Dict[str, Any]] = {}
+        self.fail_fast = conf.get("core", "fail_fast")
+
+    def execute_async(self, *args, **kwargs) -> None:
+        """
+        The method is replaced by custom trigger_task implementation.
+        """
```
Comment (potiuk) [inline airflow/executors/inprocess_executor.py]: Should we raise exception or make an assert here?

File: airflow/executors/inprocess_executor.py [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+        )
+        for _ in range(min((open_slots, len(self.queued_tasks)))):
+            key, (_, _, _, ti) = sorted_queue.pop(0)
+            self.queued_tasks.pop(key)
+            self.running.add(key)
+            self.tasks_to_run.append(ti)  # type: ignore
+
+    def end(self) -> None:
+        """
+        This is not and not async executor so we can
```
Comment (potiuk) [inline airflow/executors/inprocess_executor.py]:
```suggestion
        This is not an async executor so we can
```

File: airflow/models/taskinstance.py [hunk 14]
```diff
@@ -885,6 +887,12 @@ def _run_raw_task(
                 context = self.get_template_context()
 
                 task_copy = copy.copy(task)
+
+                if issubclass(type(task_copy), BaseSensorOperator) and \
+                        conf.get('core', 'executor') in ("InProcessExecutor", "SequentialExecutor"):
```
Comment (potiuk) [inline airflow/models/taskinstance.py]: Nice!

File: airflow/utils/log/colored_log.py [hunk 15]
```diff
@@ -30,7 +30,7 @@
 ARGS = {"attrs": ["bold"]}
 
 DEFAULT_COLORS = {
-    "DEBUG": "red",
+    "DEBUG": "green",
```
Comment (potiuk) [inline airflow/utils/log/colored_log.py]: I think this should be separated to a separate PR to community.

**Review by potiuk — CHANGES_REQUESTED** [review]
Love it. Just a few comments. I think also once we go through reviews of code we should add some information for developers in CONTRIBUTING section and most likely even in "doc" and make it part of the official documentation on how to use the InProcessExecutor for DAG debugging.

File: airflow/config_templates/default_airflow.cfg (hunk 9 above)
Comment (turbaszek) [reply airflow/config_templates/default_airflow.cfg]: How about [debug] sections simpler and can be reused?

File: airflow/executors/inprocess_executor.py (hunk 12 above)
Comment (turbaszek) [reply airflow/executors/inprocess_executor.py]: Why? This method is not used by the operator but must be implemented.

File: airflow/config_templates/default_airflow.cfg (hunk 9 above)
Comment (potiuk) [reply airflow/config_templates/default_airflow.cfg]: Yep. Why not. We can add more stuff there then in this section.

File: airflow/executors/base_executor.py (hunk 10 above)
Comment (evgenyshulman) [inline airflow/executors/base_executor.py:88]: I assume https://issues.apache.org/jira/browse/AIRFLOW-4196 is going to happen soon :)

File: airflow/executors/inprocess_executor.py [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+    def _run_task(self, ti: TaskInstance) -> bool:
+        self.log.debug("Executing task: %s", ti)
+        key = ti.key
+        try:
+            params = self.tasks_params.get(ti.key, {})
+            ti._run_raw_task(  # pylint: disable=protected-access
+                job_id=ti.job_id, **params
+            )
+            self.change_state(key, State.SUCCESS)
+            return False
```
Comment (evgenyshulman) [inline airflow/executors/inprocess_executor.py]: [minor]  False on success could be misleading.

File: airflow/executors/inprocess_executor.py [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+            priority=task_instance.task.priority_weight_total,
+            queue=task_instance.task.queue,
+        )
+        # Save params for TaskInstance._run_raw_task
+        self.tasks_params[task_instance.key] = {
+            "mark_success": mark_success,
+            "pool": pool,
+        }
+
+    def trigger_tasks(self, open_slots: int) -> None:
```
Comment (evgenyshulman) [inline airflow/executors/inprocess_executor.py]: we are copy pasting some code here with some business logic from baseexecutor, 
 I am not sure if there is another workaround here, we need TI to manage it state

let's see how it goes through code review

**Review by evgenyshulman — COMMENTED** [review]
Great Job @nuclearpinguin , let make a PR.

File: airflow/executors/base_executor.py (hunk 10 above)
Comment (turbaszek) [reply airflow/executors/base_executor.py:88]: We've already dropped support for Python 2.7 🐍

File: airflow/executors/inprocess_executor.py (hunk 16 above)
Comment (turbaszek) [reply airflow/executors/inprocess_executor.py]: Agreed!

File: airflow/executors/inprocess_executor.py (hunk 17 above)
Comment (turbaszek) [reply airflow/executors/inprocess_executor.py]: Yes it's mostly workaround for type annotations. Previously when executors where untyped it didn't matter what we queued.

**Review by potiuk — APPROVED** [review]

**Comment by turbaszek** [comment]
Closing as it was merged into master
